torch/lib/THD/base/init_methods/InitMethodTCP.cpp

```diff
@@ -206,39 +201,56 @@ InitMethod::Config initTCPMulticast(std::string group_name, rank_type world_size
   std::this_thread::sleep_for(std::chrono::milliseconds(50));
   broadcast();
 
-  std::vector<std::string> tmp_processes(processes.begin(), processes.end());
-  std::vector<std::string> arranged_processes(processes.size());
-
-  // Put already assigned ranks on their place
-  for (std::size_t i = 0; i < processes.size(); ++i) {
-    auto recv_msg = MulticastMessage(tmp_processes[i]);
-    if (recv_msg.rank >= 0) {
-      arranged_processes[recv_msg.rank] = tmp_processes[i];
-    }
+  std::vector<MulticastMessage> unpacked_msgs;
+  for (auto& msg : msgs) {
+    unpacked_msgs.emplace_back(msg);
   }
 
-  // Put rest (not assigned) in sort order
-  std::size_t last_pos = 0;
-  for (std::size_t i = 0; i < processes.size(); ++i) {
-    if (arranged_processes[i].empty()) { // not assigned place
-      for (std::size_t j = last_pos; j < processes.size(); ++j) {
-        auto recv_msg = MulticastMessage(tmp_processes[j]);
-        if (recv_msg.rank < 0) {
-          last_pos = j;
-          break;
-        }
-      }
+  return {listen_socket, packed_msg, unpacked_msgs};
+}
 
-      arranged_processes[i] = tmp_processes[last_pos];
-      last_pos++;
+InitMethod::Config initTCPMaster(struct sockaddr* addr) {
+  // TODO
+  throw std::runtime_error("non-multicast tcp initialization not supported");
+}
+
+InitMethod::Config initTCPMulticast(std::string group_name, rank_type world_size,
+                                    int assigned_rank, struct sockaddr* addr) {
+  InitMethod::Config config;
+
+  int listen_socket;
+  std::string packed_msg;
+  std::vector<MulticastMessage> msgs;
+  std::tie(listen_socket, packed_msg, msgs) = getMessages(addr, world_size,
```

Comment:
Wouldn't it be cleaner to pack the message in this function, and pass it to `getMessages`?